File: 14-TypeScript/Grey Multi-Tenant/packages/grey-elixir/lib/grey_sdk/config/options.ex
defmodule GreySdk.Config.Options do
  @moduledoc """
  Configuration options for the Grey SDK client.

  ## Options

  - `:host` - The host address of the Grey API server (required)
  - `:port` - The port to connect to (default: 443)
  - `:use_tls` - Whether to use TLS for the connection (default: true)
  - `:timeout_ms` - Request timeout in milliseconds (default: 30000)
  - `:metadata` - Additional metadata to include in requests

  ## Examples

      # Production configuration
      options = GreySdk.Config.Options.new(
        host: "api.grey.example.com",
        port: 443,
        use_tls: true
      )

      # Local development
      options = GreySdk.Config.Options.local(port: 50051)
  """

  @type t :: %__MODULE__{
          host: String.t(),
          port: non_neg_integer(),
          use_tls: boolean(),
          timeout_ms: non_neg_integer(),
          metadata: map()
        }

  @enforce_keys [:host]
  defstruct [
    :host,
    port: 443,
    use_tls: true,
    timeout_ms: 30_000,
    metadata: %{}
  ]

  @doc """
  Creates new options with the given configuration.

  ## Examples

      iex> GreySdk.Config.Options.new(host: "api.grey.example.com")
      %GreySdk.Config.Options{host: "api.grey.example.com", port: 443, use_tls: true}
  """
  @spec new(keyword()) :: t()
  def new(opts) when is_list(opts) do
    host = Keyword.fetch!(opts, :host)

    if host == "" or is_nil(host) do
      raise ArgumentError, "host is required and cannot be empty"
    end

    port = Keyword.get(opts, :port, 443)
    use_tls = Keyword.get(opts, :use_tls, true)
    timeout_ms = Keyword.get(opts, :timeout_ms, 30_000)
    metadata = Keyword.get(opts, :metadata, %{})

    validate_port!(port)
    validate_timeout!(timeout_ms)

    %__MODULE__{
      host: host,
      port: port,
      use_tls: use_tls,
      timeout_ms: timeout_ms,
      metadata: metadata
    }
  end

  @doc """
  Creates options for local development.

  ## Examples

      iex> GreySdk.Config.Options.local()
      %GreySdk.Config.Options{host: "localhost", port: 50051, use_tls: false}
  """
  @spec local(keyword()) :: t()
  def local(opts \\ []) do
    new(
      Keyword.merge(
        [host: "localhost", port: 50051, use_tls: false],
        opts
      )
    )
  end

  @doc """
  Creates options for production.

  ## Examples

      iex> GreySdk.Config.Options.production("api.grey.example.com")
      %GreySdk.Config.Options{host: "api.grey.example.com", port: 443, use_tls: true}
  """
  @spec production(String.t(), keyword()) :: t()
  def production(host, opts \\ []) do
    new(
      Keyword.merge(
        [host: host, port: 443, use_tls: true],
        opts
      )
    )
  end

  @doc """
  Returns the timeout as a duration in milliseconds.
  """
  @spec timeout(t()) :: non_neg_integer()
  def timeout(%__MODULE__{timeout_ms: timeout_ms}), do: timeout_ms

  @doc """
  Returns the full endpoint address.
  """
  @spec endpoint(t()) :: String.t()
  def endpoint(%__MODULE__{host: host, port: port}) do
    "#{host}:#{port}"
  end

  defp validate_port!(port) when is_integer(port) and port > 0 and port <= 65535, do: :ok
  defp validate_port!(port), do: raise(ArgumentError, "port must be between 1 and 65535, got: #{inspect(port)}")

  defp validate_timeout!(timeout) when is_integer(timeout) and timeout > 0, do: :ok
  defp validate_timeout!(timeout), do: raise(ArgumentError, "timeout_ms must be positive, got: #{inspect(timeout)}")
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-clojure/src/grey/sdk/options.clj
(ns grey.sdk.options
  "Grey SDK configuration options.
  
  Provides configuration management for the Grey SDK client."
  (:require [clojure.spec.alpha :as s]))

;; Specs for options
(s/def ::host string?)
(s/def ::port (s/and int? pos?))
(s/def ::use-tls boolean?)
(s/def ::timeout-ms (s/and int? pos?))
(s/def ::metadata (s/nilable map?))

(s/def ::options
  (s/keys :req-un [::host ::port ::use-tls ::timeout-ms]
          :opt-un [::metadata]))

(def default-timeout-ms
  "Default request timeout in milliseconds."
  30000)

(def default-port
  "Default port for gRPC."
  443)

(defn options
  "Creates Grey SDK options.
  
  Arguments:
    host       - Server hostname
    port       - Server port
    use-tls    - Whether to use TLS
    timeout-ms - Request timeout in milliseconds (optional, default 30000)
    metadata   - Additional metadata map (optional)"
  ([host port use-tls]
   (options host port use-tls default-timeout-ms))
  ([host port use-tls timeout-ms]
   (options host port use-tls timeout-ms nil))
  ([host port use-tls timeout-ms metadata]
   {:host host
    :port port
    :use-tls use-tls
    :timeout-ms timeout-ms
    :metadata metadata}))

(defn local
  "Creates options for local development.
  
  Arguments:
    port - Local server port (default: 8080)"
  ([]
   (local 8080))
  ([port]
   (options "localhost" port false default-timeout-ms)))

(defn production
  "Creates options for production with TLS.
  
  Arguments:
    host - Production server hostname"
  ([host]
   (production host default-port))
  ([host port]
   (options host port true default-timeout-ms)))

(defn endpoint
  "Returns the endpoint string for the options."
  [{:keys [host port]}]
  (str host ":" port))

(defn with-timeout
  "Returns options with a new timeout value."
  [opts timeout-ms]
  (assoc opts :timeout-ms timeout-ms))

(defn with-metadata
  "Returns options with additional metadata merged in."
  [opts metadata]
  (update opts :metadata merge metadata))

(defn valid?
  "Returns true if options are valid."
  [opts]
  (s/valid? ::options opts))

(defn validate
  "Validates options, throwing on invalid."
  [opts]
  (if (valid? opts)
    opts
    (throw (ex-info "Invalid options" {:explain (s/explain-str ::options opts)}))))
